rfcs/0001-meta-schedule-autotensorir.md

```diff
@@ -0,0 +1,444 @@
+    i_1, j_1, # S: the 2nd spatial tile
+    k_0,      # R: the 1st reduction tile
+    i_2, j_2, # S: the 3rd spatial tile
+    k_1,      # R: the 2nd reduction tile
+    i_3, j_3, # S: the 4th spatial tile
+)
+```
+
+In this example, the developers may tweak the tile sizes and measure the performance of the
+generated kernels to explore the opportunities of potential optimization.
+
+Generally speaking, while writing a schedule, there are often some parameters that are hard to
+determine ahead of time, for example, tile sizes, unroll steps, or which tensor intrinsics to use.
+Developers may manually enumerate possible combinations of these unknown factors, and then pick the
+best schedule according to measurement results on their device.
+
+### 3.2. Composite Schedule Rules
+
+As introduced in the previous section, in TensorIR, each schedule primitive handles only a very
+basic transformation of the IR. For example, `split` only splits a loop into two new loops. In the
+real world, the over-fine granularity of those primitives usually leads to repetitive and verbose
+scheduling code, as
+[mentioned](https://discuss.tvm.apache.org/t/rfc-tensorir-a-schedulable-ir-for-tvm/7872/43?u=junrushao1994)
+by developers in our community.
+
+To make it more convenient and modular, we allow users to register "composite schedules" that apply
+a sequence of schedule primitives according to certain analysis of the IR. For instance, a composite
+schedule may inspect a TensorIR block and decide whether we should call `compute_inline` on it.
+
+### 3.3. AutoTVM-style Design Space Description
+
+Meta schedule extends the schedule DSL with sampling instructions. When included in a schedule,
+these instructions parametrize the schedule from a single deterministic point to a space supported
+by random variables (tile size, etc.), making it possible for developers to describe the design
+space with meta schedule APIs.
+
+We can extend the matmul example above to cover all possible tilings using these sampling
+instructions:
+
+```python
+# Sample tile sizes
+i_tiles = sch.sample_perfect_tile(i, n=4)
+j_tiles = sch.sample_perfect_tile(j, n=4)
+k_tiles = sch.sample_perfect_tile(k, n=2)
+# Tile the loops according to the random variables
+i_0, i_1, i_2, i_3 = sch.split(loop=i, factors=i_tiles)
+j_0, j_1, j_2, j_3 = sch.split(loop=j, factors=j_tiles)
+k_0, k_1           = sch.split(loop=k, factors=k_tiles)
+# Organize the loops into "SSRSRS" 6-level tiles
+sch.reorder(
+    i_0, j_0, # S: the 1st spatial tile
+    i_1, j_1, # S: the 2nd spatial tile
+    k_0,      # R: the 1st reduction tile
+    i_2, j_2, # S: the 3rd spatial tile
+    k_1,      # R: the 2nd reduction tile
+    i_3, j_3, # S: the 4th spatial tile
+)
+```
+
+### 3.4. AutoScheduler-style Design Space Generation
+
+AutoScheduler (Ansor) generates schedule templates by applying their SearchRules to each stage.
+SearchRule analyzes TE and eagerly trigger schedule primitives accordingly in its internally
```

Comment:
Yeah, to clarity, definitely it is not what I intend to say, and "eagerly trigger" is not a precise term here.

The background context is:
- TE scheduling doesn't have an "eager mode" - all scheduling instructions are manipulating something called "Schedule Tree", and any transformation to TIR doesn't take in place until `tvm.lower` is called.
- To get around this (because AutoScheduler really needs to analyze the IR on the fly), AutoScheduler has to develop its own mini DSL to manipulate its own mini IR
- TensorIR doesn't have such problem, because each of its scheduling instruction is just an eager and direct transformation of the IR.

Reply:
but the context here is somewhat not quite related to meta schedule itself (more relevant to TensorIR)

Reply:
Revised:

```markdown
To generate design space, AutoScheduler (Ansor) applies a set of rules to each TE stage.
The rules analyze the TE operations and apply an internal DSL to manipulating its internal IR,
which is in the end mapped to TE schedule primitives. This process is called *sketch generation*.

Composite schedule rules work in a similar way scheduling TensorIR, as introduced in Section 3.2.
It analyzes the TensorIR and apply schedule primitives directly to TensorIR accordingly.
When applying such rules to each TensorIR block in certain order (e.g. Post-DFS Order),
it generates a sequence of schedule primitives.
If the sampling instructions are present in this sequence, 
the support of the probability space form a design space of possible schedulings.
This process is similar to the *sketch generation* phase in AutoScheduler.
```